sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala

```diff
@@ -442,3 +520,205 @@ class SQLQueryTestSuite extends QueryTest with SharedSQLContext {
+
+  // Dynamically pickles and reads into JVM side in order to mimic Python native function within
+  // Python UDFs.
+  private lazy val pythonFunc: Array[Byte] = if (isPythonAvailable) {
+    var binaryPythonFunc: Array[Byte] = null
+    withTempPath { path =>
+      Process(
+        Seq(
+          pythonExec,
+          "-c",
+          "from pyspark.sql.types import StringType; " +
+            "from pyspark import cloudpickle; " +
+            s"cloudpickle.dump((lambda x: str(x), StringType()), open('$path', 'wb'))"),
+        None,
+        "PYTHONPATH" -> s"$sourcePythonPath:$pythonPath").!!
+      binaryPythonFunc = Files.readAllBytes(path.toPath)
+    }
+    assert(binaryPythonFunc != null)
+    binaryPythonFunc
+  } else {
+    throw new RuntimeException(s"Python executable [$pythonExec] is unavailable.")
+  }
+
+  private lazy val pandasFunc: Array[Byte] = if (isPythonAvailable) {
+    var binaryPythonFunc: Array[Byte] = null
+    withTempPath { path =>
+      Process(
+        Seq(
+          pythonExec,
+          "-c",
+          "from pyspark.sql.types import StringType; " +
+            "from pyspark import cloudpickle; " +
+            s"cloudpickle.dump((lambda x: x.apply(str), StringType()), open('$path', 'wb'))"),
+        None,
+        "PYTHONPATH" -> s"$sourcePythonPath:$pythonPath").!!
+      binaryPythonFunc = Files.readAllBytes(path.toPath)
+    }
+    assert(binaryPythonFunc != null)
+    binaryPythonFunc
+  } else {
+    throw new RuntimeException(s"Python executable [$pythonExec] is unavailable.")
+  }
+
+  trait TestUDF
+
+  val workerEnv = new java.util.HashMap[String, String]()
+  workerEnv.put("PYTHONPATH", s"$sourcePythonPath:$pythonPath")
+
+  class TestPythonUDF() extends TestUDF {
+    lazy val udf = UserDefinedPythonFunction(
+      name = "udf",
+      func = PythonFunction(
+        command = pythonFunc,
+        envVars = workerEnv.clone().asInstanceOf[java.util.Map[String, String]],
+        pythonIncludes = List.empty[String].asJava,
+        pythonExec = pythonExec,
+        pythonVer = pythonVer,
+        broadcastVars = List.empty[Broadcast[PythonBroadcast]].asJava,
+        accumulator = null),
+      dataType = StringType,
+      pythonEvalType = PythonEvalType.SQL_BATCHED_UDF,
+      udfDeterministic = true)
+  }
+
+  class TestScalarPandasUDF() extends TestUDF {
+    lazy val udf = UserDefinedPythonFunction(
+      name = "udf",
+      func = PythonFunction(
+        command = pandasFunc,
+        envVars = workerEnv.clone().asInstanceOf[java.util.Map[String, String]],
+        pythonIncludes = List.empty[String].asJava,
+        pythonExec = pythonExec,
+        pythonVer = pythonVer,
+        broadcastVars = List.empty[Broadcast[PythonBroadcast]].asJava,
+        accumulator = null),
+      dataType = StringType,
+      pythonEvalType = PythonEvalType.SQL_SCALAR_PANDAS_UDF,
+      udfDeterministic = true)
+  }
+
+  class TestScalaUDF() extends TestUDF {
```

Comment:
Yea .. I was actually thinking about that too :-) so I named it from Python blabla to `IntegratedUDFTestUtils`. I believe it's fine because we can compare to the results of Scala UDF